### PR TITLE
Assets as first class concept

### PR DESCRIPTION
**Pull request by bgrundmann**
This is very much **NOT** ready to pull.  But I would like some feedback.  I originally just wanted to make the url used by font awesome configurable (as I use a more recent version personally and also to allow one to use licensed locally installed versions if one so desires).  And I didn't particularly like that currently the font awesome url is in every `base.html`.

But I ended up writing something that might just make all but `base.html` unnecessary. So if one would take what I did to the logical extremes, one could delete `base_bulma.html, base_foundation.html` and so on unnecessary.

Maybe more importantly the code would compose more nicely that is if hypothetically one would compose a style using a different icon package + bootstrap base, there would be no need to separately touch some template file somewhere.

That all said this is not all roses yet.  Maybe most importantly the way I crowbarred support for assets into `Style.py` and then bound it inside `render_root` probably shows that I haven't really understood neither tri.declarative nor how it is used in iommi.

I also haven't done the work to adjust all the styles (in fact I haven't even done the remaining, albeit trivial work) to get rid of what is left of `base_bulma.html`.

And lastly the code that gathers all the assets together in `Style.__init__ `should be changed such that the resulting order is both what one would expect and controllable by the programmer.

Thoughts?

### CODE REVIEW

**Comment by boxed** [comment]
I like it!

I think the biggest change would be that assets should be a namespace where all the components are named. So for example for fontawesome you'd have `assets__icons=....`, and then for bulma you'd have `assets__css=...`, and `assets__navburger_script`. This way the assets compose, which they don't if they are a list that can only be overwritten totally.

So to use a local bulma css file you'd make your own style: `Style(bulma, assets__css='my_local_thing'....` and use that instead. 

This would also mean that the `base` style could have `assets__select2` and since all the styles are based on `base` they'd get this automatically. This should also mean that to get rid of select2 you could do `assets__select2=None` in your style.

**Comment by bgrundmann** [comment]
Hmm I agree on a conceptual level -- I think.  At least the word Namespace sounds right.  But practically I'm staring at the "documentation" of Namespace (the source wasn't much help either).  And I'm not even sure I understand what exactly a namespace represents.  Much less the exact API.  That said its early in the morning and I'm not sure the coffee is working...  If you can give me a few hints I'll be grateful, otherwise expect this to take a little while until I understand what's going on.

```

class tri_declarative.Namespace(*dicts, **kwargs)[source]

    __call__(*args, **kwargs)[source]

        Call self as a function.

    __init__(*dicts, **kwargs)[source]

        Initialize self. See help(type(self)) for accurate signature.

    __repr__()[source]

        Return repr(self).

    __str__()[source]

        Return str(self).

    __weakref__

        list of weak references to the object (if defined)
```

And I must say I feel a little like I'm missing

**Comment by jlubcke** [comment]
The Namespace-class is the result of several years of polishing the same 30 lines of code. The only definitive blueprint is the head of Anders and me :^) I probably should try to write something with a lot of examples in the documentation on github.

**Comment by jlubcke** [comment]
(But it has stopped moving and almost always does exactly what we want I think it is the right thing)

**Comment by boxed** [comment]
The idea of namespaces is roughly that these things are roughly logically equivalent:

```python
car = Car()
car.engine.cylinders = 4

car = Car(engine__cylinders=4)

car = Car(engine=dict(cylinders=4))
```

I'm waving my hands in the air a bit vaguely now :) To make stuff like this work everything must have a name (so no lists!).

**Comment by boxed** [comment]
We've tried to write about this: 
https://docs.iommi.rocks/en/latest/architecture.html 
https://docs.iommi.rocks/en/latest/philosophy.html
https://kodare.net/2018/06/25/refinableobject-object-orientation-refined.html
https://kodare.net/2020/09/14/transparent_apis.html

As you see we're still trying to find a way to explain this in a way everyone can understand :P

**Comment by jlubcke** [comment]
The Namespace class is what does the parsing of paths with '__', and the mergeing with other namespaces.
For example:
```python
from tri_declarative import Namespace
n = Namespace(foo__bar=17)
assert n.foo.bar == 17

defaults = Namespace(foo__bar=1, foo__baz=2)
result = Namespace(defaults, n)
assert result.foo == dict(bar=17, baz=2)
```

**Comment by jlubcke** [comment]
After back-channel feedback it was suggested to use less abstract variable names:
```python
from tri_declarative import Namespace
banana = Namespace(peel__color='yellow')
assert banana.peel.color == 'yellow'

defaults = Namespace(peel__color='red', peel__thickness=1)
result = Namespace(defaults, banana)
assert result.peel == dict(color='yellow', thickness=1)
```

**Comment by jlubcke** [comment]
Another hint: The Namespace class is mostly used through the `@dispatch` decorator on the constructor.

**Comment by bgrundmann** [comment]
Thanks that is all very useful.  I did indeed try to use `@dispatch` on `Style.__init__` but I think the additional magic done on it's `kwargs` arguments for `self.config` collide with that.  And I don't yet understand what that part of the function does well enough.

Anyhow I have  to run off to a meeting but I'll have another look at it later today.

**Comment by boxed** [comment]
Me and Johan have also been discussing assets generally. At work I built a little wrapper for `render_bundle` that makes sure you can use it many times in a template but you only get that asset bundle rendered once in the html. This makes it much easier to work with as you can have stuff like assets per component and if you render 100 of that component you get the assets once, but if you have zero you don't get that asset. I think we should do something similar here. I'm going to think about this a bit :P

**Comment by bgrundmann** [comment]
Well basically if we would just give each distinct asset a unique id than it would be pretty trivial to first collect all assets exactly once in a  set / dict, inside render_root.  

So the question really becomes what should that id be.  

Options:

- string representation of the fragment (means one could even have different libraries that didn't know about each other and just happen to require the same asset and still one requires the asset just once, for that to work one has to make sure that the fragments are always generated in the same way, but some nice helpers for link to javascript, link to css (with / without the various optional arguments) and inline-script should take care of that.
 
- id of the python object representing the asset, disadvantage: now one needs to explicitely share the python objects to get unification.  Advantage it is pretty obvious what will and will not get shared.

- Some explicit id stored with the asset object.  Roughly the same advantage / disadvantage as using the python id.  One could sometimes explicitely choose the id to get sharing.

- One could also imagine designing a very explicit asset naming path and always just using one thing per path.   That path could of course be an iommi name.  But I suspect down that path hilarious bugs can be found (imagine getting a slightly different url depending on the order in which some libraries are imported...)

All in I am leaning towards the string representation unification proposal, but then you have probably build more websites than me...

**Comment by boxed** [comment]
The iommi way(tm) is to name all the parts. So it's really not one of those at all I would say! It's a user supplied name on the way in, like the names of `Field`s in a `Form` all have names. So we'd have the same for our assets. 

I'm building this now. It's going to be great! :P

**Comment by bgrundmann** [comment]
K -- I'll just step aside and watch the master at work ;-)

**Comment by boxed** [comment]
Well.. it will be great. It's just that there's a nasty bug that mutates stuff on the side during styling application that this feature is blocked on. It just so happens Johan is already working on fixing that for an unrelated reason :P

**Comment by boxed** [comment]
There we go. This is what I think it should look like but as I said it's broken right now. 

But I hope you see where I was going with the general sort of feel of the code...

**Comment by bgrundmann** [comment]
Hmm can I ask a question?  I don't understand that line in `Style.__init__` :

```python
    self.assets = setdefaults_path(
            *(base.assets for base in bases),
            assets,
        )
```

Doesn't that mean that we don't even know exactly what assets will end up being passed as __target__ to setdefaults_path (it might be one of the one in bases, or assets itself is bases is [])? And that should matter shouldn't it because one should generally know what will and will not get mutated?

I am clearly missing something what is it?

**Comment by boxed** [comment]
Staring at that section a bit I realized I was missing a `reversed`. It should look like this:

```python
        self.assets = setdefaults_path(
            *(base.assets for base in reversed(bases)),
            assets,
        )
```

It means that we take the assets of the bases and then the assets of the style object itself. So let's take a concrete example of the bulma `Style`. First there's the assets defined in `base`:

```python
    assets=dict(
        jquery=...,
        select2_js=...,
    ),

Then there's the assets defined in `bulma`:

```python
    assets=dict(
        css=...,
        js=....,
    ),
```

The end result of this will be a merged dict looking like:

```python
    assets=dict(
        jquery=...,
        select2_js=...,
        css=...,
        js=....,
    ),
```

Now you can make a new style:

```pytho
foo = Style(
    bulma,
    assets__js=None,
)
```

which just throws away the js. 

Does that make sense?

**Comment by bgrundmann** [comment]
Thanks for taking the time to write that long explanation.  Sadly I think I didn't express my confusion correctly.  Let me try again.

The part that I don't understand is that `setdefaults_path` modifies it's first argument (`__target__` by calling `dict.update` on it).

So doesn't that mean that you are modifying in your example `base.assets` to look like that merged dictionary you mentioned when the `Style.__init__` for `bulma` has run?  And isn't that undesirable?  I mean sure one is unlikely to actually use the base style anywhere but still...

**Comment by jlubcke** [comment]
The right way to do it is `setdefaults_path(Namespace(), ..., ...)`

**Comment by jlubcke** [comment]
The best way to trip up is to have an audience ;^)

**Comment by boxed** [comment]
Awesome find. Yes. I fucked up, then I blamed an unrelated bug for my fuckup :(((((

I have pushed a fix.

**Comment by boxed** [comment]
This all works super nicely now btw! Thanks!

**Comment by bgrundmann** [comment]
NP.  Glad to hear that I was confused for good reason ;-).  The ship has probably sailed as the lack of a proper type checker makes it much harder to modify function signatures after the fact.  But maybe setdefaults_path should have taken target as a required keyword argument.  Something like: `setdefaults_path(target=..., ...)`

**Comment by boxed** [comment]
It can't really, because then you can't set a member called "target":

```python
setdefaults_path(
   target=foo,
   target=bar,
)
```
...would be invalid python.

**Comment by bgrundmann** [comment]
Sigh makes sense.

**Comment by boxed** [comment]
I think I'm done in moving all the assets to this new much improved system. I would like @jlubcke to review but then I think we can merge this.

**Comment by bgrundmann** [comment]
Sweet
